presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantAggregateDistinct.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.AggregationNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.Rule;
+
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.spi.plan.AggregationNode.Aggregation.removeDistinct;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+
+/**
+ * Removes distinct from aggregates where the combination of aggregate columns and grouping variables contain a unique key.
+ * Ultimately this optimization needs to happen before the mark distinct optimization occurs.
+ * This will require moving the operations that transform away original expressions earlier in the sequence
+ * as logical property computation is designed to sit behind that transformation. For now this rule
+ * can be tested by disabling the mark distinct rule.
+ */
+public class RemoveRedundantAggregateDistinct
+        implements Rule<AggregationNode>
```

Comment:
One issue with this rule is that maybe other rules that could have benefited from the fact that this is a distinct can no longer detect that. I'm not sure if there is a way to indicate distinctness without having to actually performing distinct

Reply:
Redundant operators (DISTINCT, SORT, LIMIT, TOPk, etc) should not be applied at runtime and should be removed. The logical properties can be tested to determine "distinctness" after removal. E.g. AggNode (keyRequirement:(X,Y) , TableScanNode()) becomes TableScanNode() because TableScanNode has a key that satisfies the keyRequirement(X,Y). No information is lost when the redundant operator is removed.